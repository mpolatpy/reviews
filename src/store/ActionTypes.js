export const ReviewActionTypes = {
    UPDATE_RESPONSE: 'UPDATE_RESPONSE',
    DELETE_RESPONSE: 'DELETE_RESPONSE',
}