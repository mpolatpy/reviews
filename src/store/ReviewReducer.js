import reviewList from './reviews.json';
import { ReviewActionTypes } from './ActionTypes';

export const initialState = {
    reviews: reviewList.reduce((acc, curr) => {
        acc[curr.id] = curr;
        return acc;
    }, {})
}

export function reviewReducer(state, action) {
    switch (action.type) {
        case ReviewActionTypes.UPDATE_RESPONSE: {
            return {
                ...state,
                reviews: {
                    ...state.reviews,
                    [action.payload.id]: {
                        ...state.reviews[action.payload.id],
                        response: action.payload.response
                    }
                }
            }
        }
        case ReviewActionTypes.DELETE_RESPONSE: {
            console.log(action, state)
            return {
                ...state,
                reviews: {
                    ...state.reviews,
                    [action.payload.id]: {
                        ...state.reviews[action.payload.id],
                        response: undefined
                    }
                }
            }
        }
        default: {
            throw Error('Unknown action: ' + action.type);
        }
    }
}