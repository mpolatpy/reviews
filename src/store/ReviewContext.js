import { createContext, useContext, useReducer } from 'react';
import { initialState, reviewReducer } from './ReviewReducer';

export const ReviewContext = createContext(null);
export const ReviewDispatchContext = createContext(null);

export function ReviewProvider({ children }) {
    const [state, dispatch] = useReducer(
        reviewReducer,
        initialState
    );

    return (
        <ReviewContext.Provider value={state}>
            <ReviewDispatchContext.Provider value={dispatch}>
                {children}
            </ReviewDispatchContext.Provider>
        </ReviewContext.Provider>
    );
}

export function useReviews() {
    return useContext(ReviewContext);
}

export function useReviewsDispatch() {
    return useContext(ReviewDispatchContext);
}