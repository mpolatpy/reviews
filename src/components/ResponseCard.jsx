import React, { useState } from "react";
import MoreActionsButton from "./MoreActionsButton";
import Card from "@mui/material/Card";
import ReplyIcon from '@mui/icons-material/Reply';
import IconButton from "@mui/material/IconButton";
import Box from "@mui/material/Box";
import Tooltip from "@mui/material/Tooltip";
import TextField from '@mui/material/TextField';
import Typography from "@mui/material/Typography";
import { useReviews, useReviewsDispatch } from "../store/ReviewContext";
import { ReviewActionTypes } from "../store/ActionTypes";

const ResponseCard = ({ maxWidth, reviewId }) => {
    const dispatch = useReviewsDispatch();
    const { reviews } = useReviews();
    const review = reviews[reviewId];
    const [isEditing, setIsEditing] = useState(!review.response);
    const [responseText, setResponseText] = useState(review.response?.text ?? '');

    const menuItems = [
        {
            text: 'Edit',
            handleAction: () => setIsEditing(true)
        },
        {
            text: 'Delete',
            handleAction: () => {
                dispatch({
                    type: ReviewActionTypes.DELETE_RESPONSE,
                    payload: {
                        id: reviewId
                    }
                });

                setResponseText('');
                setIsEditing(true);
            }
        },
    ];

    const handleSaveResponse = () => {
        dispatch({
            type: ReviewActionTypes.UPDATE_RESPONSE,
            payload: {
                id: reviewId,
                response: {
                    author: 'Current User',
                    published_at: new Date().toString(),
                    text: responseText
                }
            }
        });
        setIsEditing(false);
    }

    return (
        <Card sx={{ maxWidth: maxWidth, minWidth: 300, borderRadius: '1px', p: 2 }}>
            <Box
                sx={{
                    display: 'flex',
                    flexDirection: 'row',
                }}
            >
                <Box sx={{ marginRight: '10px' }}>
                    <Tooltip title="Save Response">
                        <span>
                            <IconButton
                                size="large"
                                disabled={!isEditing || responseText.length === 0}
                                p={1}
                                onClick={handleSaveResponse}
                            >
                                <ReplyIcon fontSize="large" color="primary" />
                            </IconButton>
                        </span>
                    </Tooltip>
                </Box>
                <Box
                    sx={{
                        margin: '0 10px',
                        width: '90%',
                    }}
                >
                    {
                        isEditing ? (
                            <TextField
                                multiline
                                minRows={2}
                                sx={{ marginTop: '10px' }}
                                fullWidth
                                value={responseText}
                                size="small"
                                placeholder="Enter your response here"
                                onChange={(e) => setResponseText(e.target.value)}
                            />
                        ) : (
                            <Typography sx={{ marginTop: '20px' }} variant="body2">
                                {review.response?.text ?? ''}
                            </Typography>
                        )
                    }
                </Box>
                <MoreActionsButton menuItems={menuItems} />
            </Box>
            {
                review.response && (
                    <Box
                        sx={{
                            marginTop: '30px',
                            display: 'flex',
                            flexDirection: 'row',
                        }}
                    >
                        <Typography variant='body2'>
                            {review.response.author}
                        </Typography>
                        <Typography ml={3} variant='body2' sx={{ opacity: '50%' }}>
                            {new Date(review.response.published_at).toLocaleDateString()}
                        </Typography>
                    </Box>
                )
            }
        </Card>
    )
}

export default ResponseCard;