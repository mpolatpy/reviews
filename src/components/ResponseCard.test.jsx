import { render, screen, fireEvent } from '@testing-library/react';
import ResponseCard from './ResponseCard';
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import { ReviewProvider } from '../store/ReviewContext';

describe('ResponseCard component tests', () => {
    let wrapper, history, text;

    beforeEach(() => {
        history = createMemoryHistory();
        text = 'Sample Response';

        wrapper = render(
            <ReviewProvider>
                <Router history={history}>
                    <ResponseCard reviewId='5d707203b65083001e956f0a' />
                </Router>
            </ReviewProvider>
        );
    });

    test('matches the snaphot', () => {
        expect(wrapper).toMatchSnapshot();
    });


    test('save response button is disabled', () => {
        const responseButton = screen.getAllByRole('button')[0];
        expect(responseButton).toBeDisabled();
    });

    test('save response button is enabled after text input entry', () => {
        const responseButton = screen.getAllByRole('button')[0];
        const textInput = screen.getByRole('textbox');
        fireEvent.change(textInput, { target: { value: text } });

        expect(responseButton).toBeEnabled();

        fireEvent.click(responseButton);
        
        expect(screen.getByText(text)).toBeInTheDocument();
    });
})