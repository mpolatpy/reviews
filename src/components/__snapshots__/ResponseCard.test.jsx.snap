// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponseCard component tests matches the snaphot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1rrned0-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiBox-root css-1xhj18k"
        >
          <div
            class="MuiBox-root css-lvyu5j"
          >
            <span
              aria-label="Save Response"
              class=""
              data-mui-internal-clone-element="true"
            >
              <button
                class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
                disabled=""
                p="1"
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-bsqp9w-MuiSvgIcon-root"
                  data-testid="ReplyIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 9V5l-7 7 7 7v-4.1c5 0 8.5 1.6 11 5.1-1-5-4-10-11-11z"
                  />
                </svg>
              </button>
            </span>
          </div>
          <div
            class="MuiBox-root css-10o8l7o"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-2y6tlo-MuiFormControl-root-MuiTextField-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-z1o969-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  placeholder="Enter your response here"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div>
            <button
              aria-haspopup="true"
              aria-label="More Actions"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              id="more-actions-button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
                data-testid="MoreHorizIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1rrned0-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiBox-root css-1xhj18k"
      >
        <div
          class="MuiBox-root css-lvyu5j"
        >
          <span
            aria-label="Save Response"
            class=""
            data-mui-internal-clone-element="true"
          >
            <button
              class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
              disabled=""
              p="1"
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-bsqp9w-MuiSvgIcon-root"
                data-testid="ReplyIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 9V5l-7 7 7 7v-4.1c5 0 8.5 1.6 11 5.1-1-5-4-10-11-11z"
                />
              </svg>
            </button>
          </span>
        </div>
        <div
          class="MuiBox-root css-10o8l7o"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-2y6tlo-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-multiline css-z1o969-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <textarea
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                placeholder="Enter your response here"
                style="height: 0px; overflow: hidden;"
              />
              <textarea
                aria-hidden="true"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputSizeSmall css-12tl3rr-MuiInputBase-input-MuiOutlinedInput-input"
                readonly=""
                style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
                tabindex="-1"
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div>
          <button
            aria-haspopup="true"
            aria-label="More Actions"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            id="more-actions-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
              data-testid="MoreHorizIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
